### PR TITLE
parameterize en/decoder

### PR DESCRIPTION
**Pull request by cseed**
- made decoder stuff modular/composable
 - CodecSpec, stored in RVDSpec (and hence in metadata) describes the en/decoding process
 - exposed to Python (private parameter in MT and Table.write)

### CODE REVIEW

**Comment by cseed** [comment]
FYI, builds on: https://github.com/hail-is/hail/pull/2858

File: src/main/scala/is/hail/table/Table.scala [hunk 1]
```diff
@@ -1056,34 +1085,33 @@ class Table(val hc: HailContext, val ir: TableIR) {
 
   def collect(): IndexedSeq[Row] = rdd.collect()
 
-  def write(path: String, overwrite: Boolean = false) {
-    if (!path.endsWith(".kt") && !path.endsWith(".kt/"))
-      fatal(s"write path must end in '.kt', but found '$path'")
+  def write(path: String, overwrite: Boolean = false, codecSpecJSONStr: String = null) {
```
Comment (danking) [inline src/main/scala/is/hail/table/Table.scala]: This isn't a performance critical place, we generally have preferred explicit option types over relying on the nullability of reference types.

File: src/test/scala/is/hail/annotations/UnsafeSuite.scala [hunk 2]
```diff
@@ -25,26 +26,30 @@ class UnsafeSuite extends SparkSuite {
...
     val p = Prop.forAll(g) { case (t, a) =>
       t.typeCheck(a)
-      val f = t.fundamentalType
 
-      region.clear()
-      rvb.start(f)
-      rvb.addRow(t, a.asInstanceOf[Row])
-      val offset = rvb.end()
-      val ur = new UnsafeRow(t, region, offset)
+      println("loop")
```
Comment (danking) [inline src/test/scala/is/hail/annotations/UnsafeSuite.scala]: rogue println

File: src/test/scala/is/hail/annotations/UnsafeSuite.scala [hunk 3]
```diff
@@ -25,26 +26,30 @@ class UnsafeSuite extends SparkSuite {
...
+      CodecSpec.codecSpecs.foreach { codecSpec =>
+        println(JsonMethods.compact(JsonMethods.render(codecSpec.toJSON)))
```
Comment (danking) [inline src/test/scala/is/hail/annotations/UnsafeSuite.scala]: another one, though I understand why it's here. We should really have something like forAll but actually for every single element of this seq.

File: src/main/scala/is/hail/io/RowStore.scala [hunk 4]
```diff
@@ -2,12 +2,218 @@ package is.hail.io
...
+  def extract(jv: JValue): LEB128BufferSpec = new LEB128BufferSpec(BufferSpec.extract(jv \ "child"))
+}
+
+final class LEB128BufferSpec(child: BufferSpec) extends BufferSpec {
+  def buildInputBuffer(in: InputStream): InputBuffer = new LEB128InputBuffer(child.buildInputBuffer(in))
+
+  def buildOutputBuffer(out: OutputStream): OutputBuffer = new LEB128OutputBuffer(child.buildOutputBuffer(out))
+
+  def toJSON: JValue = JObject("name" -> JString("LEB128Buffer"),
+    "args" -> JObject("child" -> child.toJSON))
```
[the remark this answers is not in the document]
Comment (cseed) [reply src/main/scala/is/hail/io/RowStore.scala]: What's the argument?  I thought in the other PR you argued I shouldn't write my own parser for types but instead use a standard interchange format like JSON.  That's what I'm doing here.

[the remark this answers is not in the document]
Comment (danking) [reply src/main/scala/is/hail/io/RowStore.scala]: Sorry, I'm referring to [my comments on the PR on which this one builds](https://github.com/hail-is/hail/pull/2858/files#r166441106). Json4s has infrastructure for automatically doing this kind of boilerplate.

File: src/main/scala/is/hail/io/RowStore.scala [hunk 5]
```diff
@@ -2,12 +2,218 @@ package is.hail.io
...
+    jv \ "name" match {
+      case JString("LEB128Buffer") =>
+        LEB128BufferSpec.extract(jv \ "args")
+      case JString("BlockingBuffer") =>
+        BlockingBufferSpec.extract(jv \ "args")
+    }
+  }
+}
+
+abstract class BufferSpec extends Serializable {
```
[the remark this answers is not in the document]
Comment (cseed) [reply src/main/scala/is/hail/io/RowStore.scala]: Why?  These are only used for single inheritance so it shouldn't matter and I thought abstract classes were potentially faster than traits.  It is important these be fast, obviously.

From: http://www.artima.com/pins1ed/traits.html#12.7:

"Most Java runtimes make a virtual method invocation of a class member a faster operation than an interface method invocation."

File: src/main/scala/is/hail/table/Table.scala (hunk 1 above)
Comment (cseed) [reply src/main/scala/is/hail/table/Table.scala]: I generally have a preference away from Option, period.  py4j is already known to be painfully slow and changing this to an option just adds more round trips and I think all the option nonsense just leads to noisier and more error prone Python.  If you push, I will change.

File: src/main/scala/is/hail/io/RowStore.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (danking) [reply src/main/scala/is/hail/io/RowStore.scala]: I have not heard this before. I dug around a bit and found [an SO post](https://stackoverflow.com/questions/6839943/why-are-interface-method-invocations-slower-than-concrete-invocations) which suggests this is mostly a performance myth. They link to a [doc page for the Dalvik VM](https://developer.android.com/training/articles/perf-tips.html#Myths) but I couldn't find any information about HotSpot. HotSpot is supposed to be very good at

In general, we should use the weakest possible language form that satisfies our need. In this case it's a `trait`.

It's my opinion / my software-engineering-philosophy that we should only sacrifice engineering best practices when we have clear evidence that they impede our success.

From my understanding of JVM performance, a big factor in method call performance is the degree of [de facto polymorphism](https://wiki.openjdk.java.net/display/HotSpot/VirtualCalls). If performance when reading is disappointing we should investigate the degree of polymorphism present at runtime (using `jstat` and friends).

**Comment by cseed** [comment]
Changed.  After warming up the CPU so the fan was reading steady, they seem to profile similarly on read/count for a 2GB dataset (abs. class 18.8 vs trait 18.5 best of 3, both other runs were +/- 3s).

 > I have not heard this before.

This makes total sense from the implementation perspective.  How would you implement an abstract class/virtual method vs trait?  As far as I can tell in the general case the trait is going to require one more level of indirection.  See the second comment in the linked SO post.

**Comment by danking** [comment]
@cseed I think the object's name is being mangled by Scala. It's probably `CodecSpec$` or something similar.

**Pull request merged**